### PR TITLE
Add typing in ``pylint.reporters``

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |
| ✓   | :scroll: Docs          |

## Description

This add typing in ``pylint.reporters`` in order to add typing little by little in the code base. Necessary for easier evolution of json output for 3.0.

### CODE REVIEW

File: pylint/checkers/imports.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/imports.py:193]: In order to fix:
```
pylint/checkers/imports.py:193: error: Argument 1 to "Paragraph" has incompatible type "str"; expected "Union[List[Text], Tuple[str, ...]]"
Found 1 error in 1 file (checked 197 source files)
```

File: pylint/reporters/reports_handler_mix_in.py
Comment (DanielNoord) [inline pylint/reporters/reports_handler_mix_in.py:18]: We should probably add typing of the Dict if we are taking the effort of adding typing. I could take a look if you want me to!

Comment (Pierre-Sassoulas) [reply pylint/reporters/reports_handler_mix_in.py:18]: Sure thing !

File: pylint/reporters/reports_handler_mix_in.py
Comment (DanielNoord) [inline pylint/reporters/reports_handler_mix_in.py:27]: Based on the docstring of the only function that calls `register_report()` I think we can do `checker: IChecker`.
The `Callable` seems much harder to type and might be best handled when we get to typing `linter.reports`

File: pylint/checkers/imports.py
Comment (DanielNoord) [reply pylint/checkers/imports.py:193]: Based on `BaseLayout` I think those expected value are incorrect.

File: pylint/reporters/reports_handler_mix_in.py
Comment (Pierre-Sassoulas) [reply pylint/reporters/reports_handler_mix_in.py:27]: IChecker works :)

File: pylint/reporters/__init__.py
Comment (DanielNoord) [inline pylint/reporters/__init__.py:39]:
```suggestion
    utils.register_plugins(linter, __path__[0])  # type: ignore # Fixed in https://github.com/python/mypy/pull/9454
```

File: pylint/reporters/reports_handler_mix_in.py
Comment (Pierre-Sassoulas) [reply pylint/reporters/reports_handler_mix_in.py]: Do you mean typing ``pylint/lint/report_functions.py`` ?

File: pylint/reporters/reports_handler_mix_in.py
Comment (DanielNoord) [reply pylint/reporters/reports_handler_mix_in.py:19]:
```suggestion
        self._reports_state: Dict[str, bool] = {}
```

File: pylint/reporters/reports_handler_mix_in.py
Comment (DanielNoord) [inline pylint/reporters/reports_handler_mix_in.py:18]:
```suggestion
        self._reports: DefaultDict[IChecker, List[Tuple[str, str, Callable]] = collections.defaultdict(list)
```

File: pylint/reporters/reports_handler_mix_in.py
Comment (DanielNoord) [inline pylint/reporters/reports_handler_mix_in.py:56]: We should probably block this PR by https://github.com/PyCQA/pylint/pull/4973/
This function is also typed more specifically there.

File: pylint/reporters/reports_handler_mix_in.py
Comment (DanielNoord) [inline pylint/reporters/reports_handler_mix_in.py:72]: Same as above.

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [inline pylint/reporters/ureports/nodes.py:24]: Is `self.id` ever used? Searching for `.id` in the repository did not return anything that caught my eye. This might allow us to remove the Any typing from `nid`.

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [inline pylint/reporters/ureports/nodes.py:26]: This should be `Optional[VNode]`. But this will create a `mypy` warning as `VNode` does not have a `parent` attribute. See L79.
Using `Optional[BaseLayout]` would create problems with `append()` and `insert()`. However, I think those can be moved to the `BaseLayout` class as `Text` and `VerbatimText` should never use those methods. This would allow typing with `Optional[BaseLayout]`.

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [inline pylint/reporters/ureports/nodes.py:27]:
```suggestion
        self.children: children: List["VNode"] = []
```

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [inline pylint/reporters/ureports/nodes.py:33]:
```suggestion
    def append(self, child: "VNode") -> None:
```

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [inline pylint/reporters/ureports/nodes.py:43]:
```suggestion
    def accept(self, visitor: TextWriter, *args: Any, **kwargs: Any) -> None:
```
Do `visit_...` functions ever return anything? I don't believe so right?
Same goes for `leave()` just below here.

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [inline pylint/reporters/ureports/nodes.py:69]: This should become the `VNode` method (I think).

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [inline pylint/reporters/ureports/nodes.py:74]:
```suggestion
    def parents(self) -> List["VNode"]:
```

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [inline pylint/reporters/ureports/nodes.py:140]: https://github.com/PyCQA/pylint/pull/4973/ adds this to the `__ini__`, which might make this unneeded.

File: pylint/reporters/base_reporter.py
Comment (DanielNoord) [inline pylint/reporters/base_reporter.py:44]: Handled in https://github.com/PyCQA/pylint/pull/4973/.

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [inline pylint/reporters/ureports/nodes.py:189]:
```suggestion
        title: Optional[str] = None,
```
Never used in actual code, but we might as well restrict it for future applications. This is based on the other classes which implement `title`.

File: pylint/reporters/ureports/text_writer.py
Comment (DanielNoord) [inline pylint/reporters/ureports/text_writer.py:104]:
```suggestion
    def visit_verbatimtext(self, layout: "VerbatimText") -> None:
```

File: pylint/reporters/reports_handler_mix_in.py
Comment (DanielNoord) [reply pylint/reporters/reports_handler_mix_in.py]: Yes!

File: pylint/interfaces.py
Comment (DanielNoord) [inline pylint/interfaces.py:44]:
```suggestion
def implements(obj: "Interface", interface: Tuple[type, type]) -> bool:
```
Although I think it might be possible to restrict this even more

File: pylint/lint/report_functions.py
Comment (DanielNoord) [inline pylint/lint/report_functions.py:11]: This is all handled by https://github.com/PyCQA/pylint/pull/4973/ and is more specific.

File: pylint/lint/report_functions.py
Comment (cdce8p) [inline pylint/lint/report_functions.py:33]: This should probably be `List[str]`. https://github.com/PyCQA/pylint/pull/4973#discussion_r708572347

File: pylint/lint/report_functions.py
Comment (cdce8p) [inline pylint/lint/report_functions.py:35]: `type: ignore` not needed if it's a list.

**Review by cdce8p — COMMENTED** [review]
@DanielNoord probably knows more about that area of the code than I do. Nevertheless some small comments.

File: pylint/lint/report_functions.py
Comment (Pierre-Sassoulas) [reply pylint/lint/report_functions.py:11]: Do you think we should merge 4973 first ?

File: pylint/lint/report_functions.py
Comment (Pierre-Sassoulas) [reply pylint/lint/report_functions.py:33]:
```suggestion
    lines: List[str] = ("message id", "occurrences")
```

File: pylint/lint/report_functions.py
Comment (Pierre-Sassoulas) [reply pylint/lint/report_functions.py:35]:
```suggestion
        lines += (msg_id, str(value))
```

File: pylint/reporters/ureports/nodes.py
Comment (Pierre-Sassoulas) [reply pylint/reporters/ureports/nodes.py:24]: I did not see it either. Report id is the closest to it.

File: pylint/reporters/ureports/nodes.py
Comment (Pierre-Sassoulas) [reply pylint/reporters/ureports/nodes.py:43]: I used pyannotate so it probably return something once. Let's find out where and fix it :)

File: pylint/lint/report_functions.py
Comment (DanielNoord) [reply pylint/lint/report_functions.py:11]: Yes! Although I'm biased because it is my own PR I think it includes some typing that is more specific than what is currently in this PR.

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [reply pylint/reporters/ureports/nodes.py:24]: Probably remove then? Or do you want to ask Marc for input?

Comment (Pierre-Sassoulas) [reply pylint/reporters/ureports/nodes.py:24]: I'm going to rebase on 4973 first, so I'm waiting for it to be merged.

File: pylint/lint/report_functions.py
Comment (DanielNoord) [inline pylint/lint/report_functions.py:41]: I think this should be implicit. Marc commented on my previous PR about this, so I think he would like to leave implicit typing. See https://github.com/PyCQA/pylint/pull/4973#discussion_r703074786 and other comments in his earlier reviews.

File: pylint/lint/report_functions.py
Comment (DanielNoord) [inline pylint/lint/report_functions.py:57]: We can leave the more specific typing here

File: pylint/lint/report_functions.py
Comment (DanielNoord) [inline pylint/lint/report_functions.py:60]: I'm fine with this, although we should then remove L53.

File: pylint/reporters/reports_handler_mix_in.py
Comment (DanielNoord) [inline pylint/reporters/reports_handler_mix_in.py:91]: This can be removed as there is now a comment on L82
```suggestion
```

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [reply pylint/reporters/ureports/nodes.py:24]: Probably able to remove now 😄

File: pylint/lint/report_functions.py
Comment (DanielNoord) [reply pylint/lint/report_functions.py]: After latest commit I think we should use L53 so we can remove some of the `type: ignore` on these lines.

File: pylint/lint/report_functions.py
Comment (cdce8p) [reply pylint/lint/report_functions.py:41]: `List[str]` is just not really necessary.

Comment (Pierre-Sassoulas) [reply pylint/lint/report_functions.py]: Removed in latest rebase.

File: pylint/reporters/ureports/nodes.py
Comment (Pierre-Sassoulas) [reply pylint/reporters/ureports/nodes.py]: I don't understand, report_id seems to still be used, right ?

Comment (Pierre-Sassoulas) [reply pylint/reporters/ureports/nodes.py]: OK, there's still a little change because I chose str instead of optional[str],  but we did the same main change that was already handled in #4973.

Comment (DanielNoord) [reply pylint/reporters/ureports/nodes.py]: Yeah, I made this comment before the merge of #4973. You can leave whatever you prefer, I don't have a particular preference although I would perhaps add the comment on the same line.

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [reply pylint/reporters/ureports/nodes.py:24]:
```suggestion
        self.parent: Optional["BaseLayout"] = None
```
After #5015 this should now be possible

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [inline pylint/reporters/ureports/nodes.py:48]: Nice 😄 I don't like `**kwargs`!

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [inline pylint/reporters/ureports/nodes.py:121]: Can't we type these based on the fact that `BaseLayout` only accepts `children` which is already typed?

File: pylint/reporters/ureports/nodes.py
Comment (Pierre-Sassoulas) [inline pylint/reporters/ureports/nodes.py:129]:
```suggestion
        self.report_id: str = ""  # Used in ReportHandlerMixin.make_reports
```

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [inline pylint/reporters/ureports/nodes.py:132]: Same here. I think it can be `Union[Optional[str], Union[List["Text"], Tuple[str, ...]]]`

File: pylint/reporters/ureports/nodes.py
Comment (DanielNoord) [inline pylint/reporters/ureports/nodes.py:180]: And...

File: pylint/reporters/ureports/nodes.py
Comment (Pierre-Sassoulas) [reply pylint/reporters/ureports/nodes.py:121]: Sure, I don't understand why this part was over complicated so much.

Comment (DanielNoord) [reply pylint/reporters/ureports/nodes.py]: That seems to be a recurring question while working on typing of the codebase 😄

File: pylint/reporters/ureports/nodes.py
Comment (Pierre-Sassoulas) [reply pylint/reporters/ureports/nodes.py]: Apparently `` Iterable[Union["Text", str]]`` works for mypy, what do you think ? Are you sure we need the ``Optional[str]`` ?

Comment (DanielNoord) [reply pylint/reporters/ureports/nodes.py]: No, based on the idea of Marc that parameters should be as broad as possible using `Iterable` seems the right way forward! Good call!

Btw, I'm not sure if you just didn't have time to look at it yet but because the underlying comment has been resolved https://github.com/PyCQA/pylint/pull/5004#discussion_r709861703 never shows up for me on GitHub. Just want to make sure you see it :)

Comment (Pierre-Sassoulas) [reply pylint/reporters/ureports/nodes.py]: Yeah I missed it as it was in a resolved conversation, thank :)

**Review by DanielNoord — APPROVED** [review]

**Pull request merged**
